src/Cli/dotnet/commands/dotnet-run/RunCommand.cs

```diff
@@ -547,34 +614,78 @@ private static void ThrowUnableToRunError(ProjectInstance project)
+        private static string? DiscoverProjectFilePath(string? projectFileOrDirectoryPath, ref string[] args, out string? entryPointFilePath)
         {
+            bool emptyProjectOption = false;
             if (string.IsNullOrWhiteSpace(projectFileOrDirectoryPath))
             {
+                emptyProjectOption = true;
                 projectFileOrDirectoryPath = Directory.GetCurrentDirectory();
             }
 
-            if (Directory.Exists(projectFileOrDirectoryPath))
+            string? projectFilePath = Directory.Exists(projectFileOrDirectoryPath)
+                ? TryFindSingleProjectInDirectory(projectFileOrDirectoryPath)
+                : projectFileOrDirectoryPath;
+
+            // If no project exists in the directory and no --project was given,
+            // try to resolve an entry-point file instead.
+            entryPointFilePath = projectFilePath is null && emptyProjectOption
+                ? TryFindEntryPointFilePath(ref args)
+                : null;
+
+            if (entryPointFilePath is null && projectFilePath is null)
             {
-                projectFileOrDirectoryPath = FindSingleProjectInDirectory(projectFileOrDirectoryPath);
+                throw new GracefulException(LocalizableStrings.RunCommandExceptionNoProjects, projectFileOrDirectoryPath, "--project");
             }
-            return projectFileOrDirectoryPath;
-        }
 
-        public static string FindSingleProjectInDirectory(string directory)
-        {
-            string[] projectFiles = Directory.GetFiles(directory, "*.*proj");
+            return projectFilePath;
 
-            if (projectFiles.Length == 0)
+            static string? TryFindSingleProjectInDirectory(string directory)
             {
-                throw new GracefulException(LocalizableStrings.RunCommandExceptionNoProjects, directory, "--project");
+                string[] projectFiles = Directory.GetFiles(directory, "*.*proj");
+
+                if (projectFiles.Length == 0)
+                {
+                    return null;
+                }
+
+                if (projectFiles.Length > 1)
+                {
+                    throw new GracefulException(LocalizableStrings.RunCommandExceptionMultipleProjects, directory);
+                }
+
+                return projectFiles[0];
             }
-            else if (projectFiles.Length > 1)
+
+            static string? TryFindEntryPointFilePath(ref string[] args)
             {
-                throw new GracefulException(LocalizableStrings.RunCommandExceptionMultipleProjects, directory);
+                if (args is not [{ } arg, ..] ||
+                    !arg.EndsWith(".cs", StringComparison.OrdinalIgnoreCase) ||
+                    !File.Exists(arg))
+                {
+                    return null;
+                }
+
+                if (!HasTopLevelStatements(arg))
+                {
+                    throw new GracefulException(LocalizableStrings.NoTopLevelStatements, arg);
```

Comment:
In theory, we could remove this restriction now and let the compiler report errors if an entrypoint is missing in the compilation (consisting of the sole input file). However, if we decide to add the support for multiple files back, we would likely want this restriction again. So I think it's better to keep the restriction (to avoid a breaking change later) and relax in future if necessary.

Reply:
I'd personally be mildly in favor of removing it mostly because I think top level statements were a bad idea in the first place. (I'm not sure how to do this, but a different check that verifies that there is an entrypoint in this file would be perfect in my view.) On the other hand, I do think your logic is sound, so I think it's ok to keep this.

Reply:
> I'm not sure how to do this, but a different check that verifies that there is an entrypoint in this file would be perfect in my view.

Compiler would do this check automatically for us because we set OutputType=Exe.